File: dotnet/ApiRequest.cs
// Copyright 2026, Pulumi Corporation.  All rights reserved.

using System;
using System.Collections;
using System.Collections.Generic;
using System.Globalization;
using System.Reflection;
using System.Runtime.Serialization;

namespace Pulumi.Cloud.Sdk
{
    /// <summary>
    /// Mutable carrier populated fluently by a generated <c>*Api</c> method before
    /// it hands the request to <see cref="ApiClient"/>. Holds the HTTP verb, the
    /// templated resource path (<c>/api/.../{param}</c>), the categorized
    /// parameters, the request body, and the negotiated media types.
    /// </summary>
    public sealed class ApiRequest
    {
        public string Method { get; }
        public string ResourcePath { get; }
        public Dictionary<string, string> PathParams { get; } = new Dictionary<string, string>();
        public Dictionary<string, List<string>> QueryParams { get; } = new Dictionary<string, List<string>>();
        public List<string> ConsumesList { get; } = new List<string>();
        public List<string> ProducesList { get; } = new List<string>();
        public object BodyValue { get; private set; }
        public bool HasBody { get; private set; }

        public ApiRequest(string method, string resourcePath)
        {
            Method = method;
            ResourcePath = resourcePath;
        }

        /// <summary>Bind a <c>{name}</c> path segment. Null values are ignored.</summary>
        public ApiRequest PathParam(string name, object value)
        {
            if (value != null)
            {
                PathParams[name] = Stringify(value);
            }
            return this;
        }

        /// <summary>
        /// Append a query parameter. Null values are dropped; collections and
        /// arrays are emitted as repeated keys (<c>k=a&amp;k=b</c>); booleans render
        /// lowercase, enums via their wire value, and dates as ISO-8601.
        /// </summary>
        public ApiRequest QueryParam(string name, object value)
        {
            if (value == null)
            {
                return this;
            }

            if (value is string s)
            {
                AddQuery(name, s);
            }
            else if (value is IEnumerable e)
            {
                foreach (var item in e)
                {
                    AddQuery(name, Stringify(item));
                }
            }
            else
            {
                AddQuery(name, Stringify(value));
            }
            return this;
        }

        public ApiRequest Body(object body)
        {
            BodyValue = body;
            HasBody = true;
            return this;
        }

        public ApiRequest Consumes(string mediaType)
        {
            ConsumesList.Add(mediaType);
            return this;
        }

        public ApiRequest Produces(string mediaType)
        {
            ProducesList.Add(mediaType);
            return this;
        }

        private void AddQuery(string name, string value)
        {
            if (value == null)
            {
                return;
            }
            if (!QueryParams.TryGetValue(name, out var list))
            {
                list = new List<string>();
                QueryParams[name] = list;
            }
            list.Add(value);
        }

        // Render a scalar value the way the reference clients do: booleans as
        // lowercase literals, DateTimeOffset as ISO-8601, string enums via their
        // [EnumMember] wire value (numeric for int-backed enums), everything else
        // via its invariant string form.
        private static string Stringify(object value)
        {
            switch (value)
            {
                case null:
                    return null;
                case bool b:
                    return b ? "true" : "false";
                case string s:
                    return s;
                case DateTimeOffset dto:
                    return dto.ToString("o", CultureInfo.InvariantCulture);
                case DateTime dt:
                    return dt.ToString("o", CultureInfo.InvariantCulture);
                case Enum e:
                    return EnumWireValue(e);
                default:
                    return Convert.ToString(value, CultureInfo.InvariantCulture);
            }
        }

        private static string EnumWireValue(Enum e)
        {
            var members = e.GetType().GetMember(e.ToString());
            if (members.Length > 0)
            {
                var attr = members[0].GetCustomAttribute<EnumMemberAttribute>();
                if (attr?.Value != null)
                {
                    return attr.Value;
                }
            }
            // int/long-backed enums carry no [EnumMember]; render their numeric value.
            return Convert.ToString(Convert.ToInt64(e), CultureInfo.InvariantCulture);
        }
    }
}

File: java/src/main/java/com/pulumi/cloud/sdk/ApiRequest.java
// Copyright 2026, Pulumi Corporation.  All rights reserved.

package com.pulumi.cloud.sdk;

import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Mutable carrier populated by a generated {@code *Api} method before it hands
 * the request to {@link ApiClient}. Holds the HTTP verb, the templated resource
 * path ({@code /api/.../{param}}), the categorized parameters, the request body,
 * and the negotiated media types.
 */
public final class ApiRequest {
    final String method;
    final String resourcePath;
    final Map<String, String> pathParams = new LinkedHashMap<>();
    final Map<String, List<String>> queryParams = new LinkedHashMap<>();
    final List<String> consumes = new ArrayList<>();
    final List<String> produces = new ArrayList<>();
    Object body;
    boolean hasBody;

    public ApiRequest(String method, String resourcePath) {
        this.method = method;
        this.resourcePath = resourcePath;
    }

    /** Bind a {@code {name}} path segment. Null values are ignored. */
    public ApiRequest pathParam(String name, Object value) {
        if (value != null) {
            pathParams.put(name, stringify(value));
        }
        return this;
    }

    /**
     * Append a query parameter. Null values are dropped; collections and arrays
     * are emitted as repeated keys ({@code k=a&k=b}); booleans render lowercase,
     * enums via their wire value, and dates as ISO-8601 — matching the reference
     * clients.
     */
    public ApiRequest queryParam(String name, Object value) {
        if (value == null) {
            return this;
        }
        if (value instanceof Collection<?>) {
            for (Object item : (Collection<?>) value) {
                addQuery(name, item);
            }
        } else if (value instanceof Object[]) {
            for (Object item : (Object[]) value) {
                addQuery(name, item);
            }
        } else {
            addQuery(name, value);
        }
        return this;
    }

    public ApiRequest body(Object body) {
        this.body = body;
        this.hasBody = true;
        return this;
    }

    public ApiRequest consumes(String mediaType) {
        consumes.add(mediaType);
        return this;
    }

    public ApiRequest produces(String mediaType) {
        produces.add(mediaType);
        return this;
    }

    private void addQuery(String name, Object value) {
        if (value == null) {
            return;
        }
        queryParams.computeIfAbsent(name, k -> new ArrayList<>()).add(stringify(value));
    }

    // String.valueOf covers all wire renderings we need: Boolean -> "true"/"false",
    // enum -> its value() (via the generated toString), ZonedDateTime -> ISO-8601,
    // numbers and Strings as-is.
    private static String stringify(Object value) {
        return String.valueOf(value);
    }
}
